man/rcvstore.man
.TH RCVSTORE %manext1% 2013-02-19 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
rcvstore \- asynchronously incorporate mail into an nmh folder
.SH SYNOPSIS
.HP 5
.na
.B %nmhlibexecdir%/rcvstore
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RB [ \-create " | " \-nocreate ]
.RB [ \-unseen " | " \-nounseen ]
.RB [ \-zero " | " \-nozero ]
.RB [ \-sequence
.I name
\&...]
.RB [ \-public " | " \-nopublic ]
.ad
.SH DESCRIPTION
.B rcvstore
incorporates a message from the standard input into an
.B nmh
folder.  This command is typically used in conjunction with
mail filtering programs such as
.B slocal
and
.BR procmail ,
to filter your mail into different folders.
.PP
You may specify which folder to use with
.IR +folder .
If no folder is specified,
.B rcvstore
will use the folder given by a non-empty \*(lqInbox:\*(rq entry in
the user's profile, else it will use the folder named \*(lqinbox\*(rq.
.PP
If the switch
.B \-create
is given (it is the default) and if the specified
(or default) folder does not exist, then it will be created.  You may
disable this with the
.B \-nocreate
option.  In this case
.B rcvstore
will exit if the specified folder does not exist.
.PP
When the new message is incorporated into the folder, it is assigned
the next highest number for that folder.
.PP
.B rcvstore
will incorporate anything except zero length messages into the user's
.B nmh
folder.  It will not change the message in any way.
.PP
If the user's profile contains a
.RI \*(lq "Msg\-Protect: nnn" \*(rq
entry, it will be used as the mode to use, in octal, on the newly created
message, otherwise the
.B nmh
default mode of 0600, see
.IR chmod (1),
will be used (on filesystems that support it).
For all subsequent operations on this message, this initially assigned
mode will be preserved.
.PP
If the switch
.B \-unseen
is given (it is on by default), and if the profile entry
\*(lqUnseen\-Sequence\*(rq is present and non-empty, then
.B rcvstore
will add the newly incorporated message to each sequence
named by this profile entry.  You may use the switch
.B \-nounseen
to disable this.  These sequences will not be zeroed by
.B rcvstore
prior to adding the new message.
.PP
Furthermore, the incoming messages may be added to additional sequences
as they arrive by the use of the
.B \-sequence
switch.  As with the commands
.B pick
and
.BR mark ,
you may also use the switches
.B \-zero
and
.B \-nozero
to specify whether to zero old sequences or not.
Similarly, use of the
.B \-public
and
.B \-nopublic
switches may be used to force these sequences to be public or
private sequences.
.SS Locking and \-unseen
If you use the \*(lqUnseen-Sequence\*(rq profile entry, rcvstore could
try to read and update its sequence state while another
.B nmh
process is also trying to do so.  This can cause the sequence state to
lose track.  To avoid this, just between asynchronous invocations of
.BR rcvstore ,
do not use it without an external locking mechanism, for example, a
.IR procmailrc (5)
local lockfile, if you use the \*(lqUnseen-Sequence\*(rq profile entry.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Folder\-Protect:~^To set mode when creating a new folder
^Inbox:~^To find the default inbox
^Msg\-Protect:~^To set mode when creating a new message
^Unseen\-Sequence:~^To name sequences denoting unseen messages
.fi
.SH "SEE ALSO"
.IR mh\-sequence (5),
.IR procmailex (5),
.IR procmailrc (5),
.IR rcvdist (1),
.IR rcvpack (1),
.IR rcvtty (1)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to \*(lqInbox\*(rq profile entry"
.RB ` \-create '
.RB ` \-unseen '
.RB ` \-nozero '
.fi
.SH CONTEXT
No context changes will be attempted, with the exception of
sequence manipulation.


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/rcvstore.c
/* rcvstore.c -- asynchronously add mail to a folder
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <fcntl.h>
#include <h/signals.h>
#include <h/mts.h>

#define RCVSTORE_SWITCHES \
    X("create", 0, CRETSW) \
    X("nocreate", 0, NCRETSW) \
    X("unseen", 0, UNSEENSW) \
    X("nounseen", 0, NUNSEENSW) \
    X("public", 0, PUBSW) \
    X("nopublic", 0, NPUBSW) \
    X("zero", 0, ZEROSW) \
    X("nozero", 0, NZEROSW) \
    X("sequence name", 0, SEQSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(RCVSTORE);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(RCVSTORE, switches);
#undef X


/*
 * name of temporary file to store incoming message
 */
static char *tmpfilenam = NULL;

static void unlink_done(int) NORETURN;

int
main (int argc, char **argv)
{
    int publicsw = -1, zerosw = 0;
    int create = 1, unseensw = 1;
    int fd, msgnum;
    size_t seqp = 0;
    char *cp, *maildir, *folder = NULL, buf[BUFSIZ];
    char **argp, **arguments;
    svector_t seqs = svector_create (0);
    struct msgs *mp;
    struct stat st;

    if (nmh_init(argv[0], 2)) { return 1; }

    done=unlink_done;

    mts_init ();
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /* parse arguments */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [switches]",
			  invo_name);
		print_help (buf, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case SEQSW: 
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument name to %s", argp[-2]);

		svector_push_back (seqs, cp);
		seqp++;
		continue;

	    case UNSEENSW:
		unseensw = 1;
		continue;
	    case NUNSEENSW:
		unseensw = 0;
		continue;

	    case PUBSW: 
		publicsw = 1;
		continue;
	    case NPUBSW: 
		publicsw = 0;
		continue;

	    case ZEROSW: 
		zerosw++;
		continue;
	    case NZEROSW: 
		zerosw = 0;
		continue;

	    case CRETSW: 
		create++;
		continue;
	    case NCRETSW: 
		create = 0;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = pluspath (cp);
	} else {
	    adios (NULL, "usage: %s [+folder] [switches]", invo_name);
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /* if no folder is given, use default folder */
    if (!folder)
	folder = getfolder (0);
    maildir = m_maildir (folder);

    /* check if folder exists */
    if (stat (maildir, &st) == NOTOK) {
	if (errno != ENOENT)
	    adios (maildir, "error on folder");
	if (!create)
	    adios (NULL, "folder %s doesn't exist", maildir);
	if (!makedir (maildir))
	    adios (NULL, "unable to create folder %s", maildir);
    }

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* ignore a few signals */
    SIGNAL (SIGHUP, SIG_IGN);
    SIGNAL (SIGINT, SIG_IGN);
    SIGNAL (SIGQUIT, SIG_IGN);
    SIGNAL (SIGTERM, SIG_IGN);

    /* create a temporary file */
    tmpfilenam = m_mktemp (invo_name, &fd, NULL);
    if (tmpfilenam == NULL) {
	adios(NULL, "unable to create temporary file in %s", get_temp_dir());
    }
    chmod (tmpfilenam, m_gmprot());

    /* copy the message from stdin into temp file */
    cpydata (fileno (stdin), fd, "standard input", tmpfilenam);

    if (fstat (fd, &st) == NOTOK) {
	(void) m_unlink (tmpfilenam);
	adios (tmpfilenam, "unable to fstat");
    }
    if (close (fd) == NOTOK)
	adios (tmpfilenam, "error closing");

    /* don't add file if it is empty */
    if (st.st_size == 0) {
	(void) m_unlink (tmpfilenam);
	inform("empty file");
	done (0);
    }

    /*
     * read folder and create message structure
     */
    if (!(mp = folder_read (folder, 1)))
	adios (NULL, "unable to read folder %s", folder);

    /*
     * Link message into folder, and possibly add
     * to the Unseen-Sequence's.
     */
    if ((msgnum = folder_addmsg (&mp, tmpfilenam, 0, unseensw, 0, 0, NULL)) == -1)
	done (1);

    /*
     * Add the message to any extra sequences
     * that have been specified.
     */
    if (seqp) {
	/* The only reason that seqp was checked to be non-zero is in
	   case a -nosequence switch is added. */
	for (seqp = 0; seqp < svector_size (seqs); seqp++) {
	    if (!seq_addmsg (mp, svector_at (seqs, seqp), msgnum, publicsw,
			     zerosw))
		done (1);
	}
    }

    svector_free (seqs);
    seq_setunseen (mp, 0);	/* synchronize any Unseen-Sequence's      */
    seq_save (mp);		/* synchronize and save message sequences */
    folder_free (mp);		/* free folder/message structure          */

    context_save ();		/* save the global context file           */
    (void) m_unlink (tmpfilenam); /* remove temporary file                  */
    tmpfilenam = NULL;

    done (0);
    return 1;
}

/*
 * Clean up and exit
 */
static void
unlink_done(int status)
{
    if (tmpfilenam && *tmpfilenam)
	(void) m_unlink (tmpfilenam);
    exit (status);
}
